src/client/wire_protocol/operations.rs

```diff
@@ -0,0 +1,309 @@
+                   return_field_selector: &Option<BSON>) -> Result<(), String> {
+        let _ = match header.write(buffer) {
+            Ok(_) => (),
+            Err(e) => return Err(e)
+        };
+
+        let _ = match buffer.write_i32::<LittleEndian>(flags) {
+            Ok(_) => (),
+            Err(_) => return Err("Unable to write flags".to_owned())
+        };
+
+        for byte in full_collection_name.bytes() {
+            let _byte_reponse = match buffer.write_u8(byte) {
+                Ok(_) => (),
+                Err(_) => return Err("Unable to write full_collection_name".to_owned())
+            };
+        }
+
+        // Writes the null terminator for the collection name string.
+        let _ = match buffer.write_u8(0) {
+            Ok(_) => (),
+            Err(_) => return Err("Unable to write full_collection_name".to_owned())
+        };
+
+        let _ = match buffer.write_i32::<LittleEndian>(number_to_skip) {
+            Ok(_) => (),
+            Err(_) => return Err("Unable to write number_to_skip".to_owned())
+        };
+
+        let _ = match buffer.write_i32::<LittleEndian>(number_to_return) {
+            Ok(_) => (),
+            Err(_) => return Err("Unable to write number_to_return".to_owned())
+        };
+
+        let _ = match Message::write_bson_document(buffer, query) {
+            Ok(_) => (),
+            Err(s) => return Err(format!("Unable to write query: {}", s))
+        };
+
+        let _ = match return_field_selector {
+            &Some(ref bson) => match Message::write_bson_document(buffer, bson) {
+                Ok(_) => (),
+                Err(s) => {
+                    let str = format!("Unable to write return_field_selector: {}", s);
+
+                    return Err(str)
+                }
+            },
+            &None => ()
+        };
+
+        // match buffer.write_u8(0) {
+        //     Ok(_) => (),
+        //     Err(_) => return Err("Unable to write bson terminator".to_owned())
+        // };
+
+
+        let _ = buffer.flush();
+
+        Ok(())
+    }
+
+    /// Attemps to write a serialized message to a buffer.
+    ///
+    /// # Arguments
+    ///
+    /// `buffer` - The buffer to write to.
+    ///
+    /// # Return value
+    ///
+    /// Returns nothing on success, or an error string on failure.
+    pub fn write(&self, buffer: &mut Write) -> Result<(), String> {
+        match self {
+            /// Only the server should sent replies
+            &Message::OpReply {..} => Err("OP_REPLY should not be sent by the client".to_owned()),
+            &Message::OpQuery {
+                header: ref h,
+                flags: f,
+                full_collection_name: ref fcn,
+                number_to_skip: ns,
+                number_to_return: nr,
+                query: ref q,
+                return_field_selector: ref rfs
+            } => Message::write_query(buffer, &h, f, &fcn, ns, nr, &q, &rfs)
```

Comment:
I definitely agree that it's not optimal, but since OpQuery is an enum variant, `write_query()` would have to repeat the pattern match on `self` to be able to access the fields, which I think is probably worse.